### PR TITLE
Verify return-path for incoming messages

### LINKED ISSUE #41: Verify return-path for incoming messages
Following conditions should be met for verification to pass:
* rDNS domain of source server's IP should be equal to the hostname value presented in EHLO/HELO command
* Hostname domain presented in EHLO/HELO command should resolve to source server's IP address
* Domain in MAIL FROM should have MX record pointing to the source server

These checks are built on the assumption that we have a DNSSEC-enabled resolver and the source server does have DNSSEC enabled.

### PR DESCRIPTION
**Pull request by NamedKitten**
This should implement all the pipeline steps from issue #41 correctly.

Closes #41.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -0,0 +1,53 @@
+github.com/GehirnInc/crypt v0.0.0-20190301055215-6c0105aabd46/go.mod h1:kC29dT1vFpj7py2OvG1khBdQpo3kInWP+6QipLbdngo=
```
Comment (foxcpp) [inline go.sum]: Remove go.sum file.
Hint: You can add it to .git/info/exclude so you will not accidentally commit it again.

@emersion, or perhaps we should allow it now?

File: smtppipeline.go [hunk 2]
```diff
@@ -19,6 +19,30 @@ type SMTPPipelineStep interface {
...
+func (verifySourceHostnameStep) Pass(ctx *module.DeliveryContext, _ io.Reader) (io.Reader, bool, error) {
+	ipAddr, ok := ctx.SrcAddr.(*net.IPAddr)
+	if !ok {
+		return nil, true, nil
+	}
+
+	srcIPs, err := net.LookupIP(ctx.SrcHostname)
+	if err != nil {
+		ctx.Ctx["src_hostname_check"] = false
+		return nil, false, errors.New("Source hostname does not resolve")
```
Comment (foxcpp) [inline smtppipeline.go]: Nitpick: [Error messages should not be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

File: smtppipeline.go [hunk 3]
```diff
@@ -19,6 +19,30 @@ type SMTPPipelineStep interface {
...
+	}
+
+	for _, ip := range srcIPs {
+		if ipAddr.IP.Equal(ip) {
+			ctx.Ctx["src_hostname_check"] = true
+			return nil, true, nil
+		}
+	}
+	ctx.Ctx["src_hostname_check"] = false
+	return nil, false, errors.New("Source hostname does not resolve to source address")
```
Comment (foxcpp) [inline smtppipeline.go]: Same here.

File: smtppipeline.go [hunk 4]
```diff
@@ -281,6 +305,8 @@ func StepFromCfg(node config.Node) (SMTPPipelineStep, error) {
 		return stopStep{}, nil
 	case "require_auth":
 		return requireAuthStep{}, nil
+	case "require_source_hostname":
```
Comment (foxcpp) [inline smtppipeline.go]: Add documentation for the pipeline step to CONFIG_REFERENCE.md.

File: go.sum (hunk 1 above)
Comment (foxcpp) [reply go.sum]: Oops, didn't notice that you removed it in the next commit.

Comment (emersion) [reply go.sum]: I'm fine with allowing it. (In any case not allowing it *and* not putting it in `.gitignore` is silly)

Comment (emersion) [reply go.sum]: Should not be done in this PR though, it's unrelated

**Comment by foxcpp** [comment]
Oops. That was a bad idea. I will fix it!

**Comment by foxcpp** [comment]
NamedKitten's laptop broke so he can't finish it. I took care of it. I wonder who should review it now... @emersion?

**Comment by emersion** [comment]
Thanks!

**Pull request merged**